### ISSUE TITLE
cbc ciphers seem to be broken

### ISSUE THREAD

**Issue by JSomerstone** [phpseclib/phpseclib#938]
Using phpseclib version 2.0.1. Disabled _all_ other encryption methdos than those using cbc.

Issue seems to be related to openssl_encrypt's initialization vector.
`PHP Warning:  openssl_encrypt(): Using an empty Initialization Vector (iv) is potentially insecure and not recommended in /tmp/matador-tst/vendor/phpseclib/phpseclib/phpseclib/Crypt/Base.php on line 1013
PHP Notice:  Invalid HMAC in /tmp/matador-tst/vendor/phpseclib/phpseclib/phpseclib/Net/SSH2.php on line 2947`

Logs indicate that key exchange succeeds but response to "ssh-userauth" fails.
...
> -> NET_SSH2_MSG_NEWKEYS (since last: 0.001, network: 0s)
>                                                  
> <- NET_SSH2_MSG_NEWKEYS (since last: 0, network: 0s)
> 
> -> NET_SSH2_MSG_SERVICE_REQUEST (since last: 0.001, network: 0s)
> 00000000  00:00:00:0c:73:73:68:2d:75:73:65:72:61:75:74:68  ....ssh-userauth
> !! NET_SSH2_MSG_SERVICE_ACCEPT
> 00000000  00:00:46:a4:b4:cd:c9:81:c9:d2:65:72:61:75:74:68  ..F.......erauth
(the last row is debug-logging written when HMAC check fails)

**Comment by terrafrost** [comment]
First, I apologize for the delay in responding to this. This has been a particularly busy week at work.

That said, I'm guessing you're running PHP 5.3? I was able to reproduce the "Invalid HMAC" error on 5.3.27 but not the "empty Initialization Vector (iv)" error.

The fix for the "Invalid HMAC" error is as follows:

https://github.com/terrafrost/phpseclib/commit/bfd467532bf75c03960a5454cb734bba61c96406

I added a unit test demonstrating the fix as well:

https://github.com/terrafrost/phpseclib/commit/b972a4b36de9c059eb0a42e8639fe2a1cade52dd

As for the "empty Initialization Vector (iv)"...  I was able to reproduce in 5.3.3. Basically, it seems like that's a bug that later versions corrected. ie. for ECB mode an initialization vector is /not/ needed. The only reason I'm utilizing ECB mode in 5.3 is because you couldn't disable padding in 5.3 so I'm using trickery to remove the padding.

I could just do `str_repeat("\0", $this->block_size)` as an extra parameter for that call of `openssl_decrypt` but I'd also like to add an explanation in the comments as to why I'm doing that when superficially it should not be necessary and rather than brute force through 5.3 versions until I find the one that fixes it (since the changelog doesn't appear to offer any insight)...  I think knowing the version you're using would help.

**Comment by terrafrost** [comment]
The fix has been merged into the 1.0, 2.0 and master branches.

As for the IV error... I'm only able to reproduce the IV error on PHP 5.3.3. Without the IV explicitly defined or with it set to `''` I get the following:

```
Warning: openssl_encrypt(): Using an empty Initialization Vector (iv) is potentially insecure and not recommended
```

But if you define the IV I get this error:

```
Warning: openssl_encrypt(): IV passed is 16 bytes long which is longer than the 0 expected by selected cipher, truncating in ...
```

So basically, the only way to fix the issue is to use the error suppression operator and that would incur a performance penalty.

Sample code:

```php
openssl_encrypt('', 'aes-128-ecb', str_repeat('x', 16), true, '');
openssl_encrypt('', 'aes-128-ecb', str_repeat('x', 16), true, str_repeat('y', 16));
```

Basically, I think the best fix for this is just going to be to adjust the error_reporting level.

So I guess half the errors (the more serious half) have been fixed and for the half that hasn't been fixed there's a workaround that can be done.
